SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Storage/StorageAdapter.cs

```diff
@@ -1,158 +1,193 @@
+        {
+            var result = ExecuteSelectionSqlQuery($"SELECT COUNT(*) FROM {tableName} WHERE {columnName} = ?;", new[] { value });
+            return (int)(long)(result.FirstOrDefault()?.FirstOrDefault() ?? 0L);
+        }
+
+        public IList<object[]> Select(string tableName, string columnName, object value, string excludeColumnName, object[] excludeValues, int? limit = null)
+        {
+            var whereClause = $"{columnName} = ?";
+            var args = new List<object> { value };
+            if (excludeValues?.Length > 0)
             {
-                throw ToStorageException(e);
+                whereClause += $" AND {excludeColumnName} NOT IN ({string.Join(",", Enumerable.Repeat("?", excludeValues.Length))})";
+                args.AddRange(excludeValues);
             }
+            var limitClause = limit != null ? $" LIMIT {limit}" : string.Empty;
+            var query = $"SELECT * FROM {tableName} WHERE {whereClause}{limitClause};";
+            return ExecuteSelectionSqlQuery(query, args);
         }
 
-        public Task InitializeStorageAsync()
+        public void Insert(string tableName, string[] columnNames, ICollection<object[]> values)
         {
-            return Task.Run(() =>
-            {
-                // Create the directory in case it does not exist.
-                if (_databaseDirectory != null)
-                {
-                    try
-                    {
-                        _databaseDirectory.Create();
-                    }
-                    catch (Exception e)
-                    {
-                        throw new StorageException("Cannot initialize SQLite library.", e);
-                    }
-                }
-
-                // In SQLite-net 1.5.231 constructor parameters were changed.
-                // Using reflection to accept newer library version.
-                _dbConnection = (SQLiteAsyncConnection)typeof(SQLiteAsyncConnection)
-                    .GetConstructor(new[] { typeof(string), typeof(bool) })
-                    ?.Invoke(new object[] { _databasePath, true });
-                if (_dbConnection == null)
-                {
-                    _dbConnection = (SQLiteAsyncConnection)typeof(SQLiteAsyncConnection)
-                        .GetConstructor(new[] { typeof(string), typeof(bool), typeof(object) })
-                        ?.Invoke(new object[] { _databasePath, true, null });
-                }
-                if (_dbConnection == null)
-                {
-                    throw new StorageException("Cannot initialize SQLite library.");
-                }
-            });
+            var columnsClause = string.Join(",", columnNames);
+            var valueClause = string.Join(",", Enumerable.Repeat("?", values.First().Length));
+            var valuesClause = string.Join(",", Enumerable.Repeat($"({valueClause})", values.Count));
+            var valuesArray = values.SelectMany(i => i).ToArray();
```

Comment:
Can you explain how does `values.SelectMany(i => i)` work? I don't understand how it would work.

Reply:
`[[1,2],[3,4],[5,6]]` -> `[1,2,3,4,5,6]`

Reply:
This is really cool, thanks.